src/jacobi/_jacobi.py

```diff
@@ -1 +1,261 @@
+        fd = [np.reshape(r.copy(), -1)]
+
+        if jac is None:  # first iteration
+            jac = np.zeros(r.shape + x.shape, dtype=r.dtype)
+            err = np.zeros(r.shape + x.shape, dtype=r.dtype)
+            if diagnostic is not None:
+                diagnostic["call"] = np.zeros((r.size, x.size), dtype=np.uint8)
+
+        if diagnostic is not None:
+            diagnostic["method"][k] = md
+            diagnostic["call"][:, k] = 2 if md == 0 else 3
+
+        for i in range(1, len(h)):
+            fdi = _derive(md, f0, fn, x, kx, h[i], args)
+            fd.append(np.reshape(fdi, -1) if i == 1 else fdi[todo])
+            if diagnostic is not None:
+                diagnostic["call"][todo, k] += 2
+                diagnostic["iteration"][k] += 1
+
+            # polynomial fit with one extra degree of freedom;
+            # use latest maxgrad + 1 data points
+            grad = min(i - 1, maxgrad)
+            start = i - (grad + 1)
+            stop = i + 1
+            q, c = np.polyfit(
+                h[start:stop] ** 2, fd[start:], grad, rcond=None, cov=True
+            )
+            ri = q[-1]
+            # pulls have roughly unit variance, however,
+            # the pull distribution is not gaussian and looks
+            # more like student's t
+            rei = c[-1, -1] ** 0.5
+
+            # update estimates that have smaller estimated error
+            sub_todo = rei < re[todo]
+            todo1 = todo.copy()
+            todo[todo1] = sub_todo
+            r[todo] = ri[sub_todo]
+            re[todo] = rei[sub_todo]
+
+            # do not improve estimates further which meet the tolerance
+            if rtol > 0:
+                sub_todo &= rei > rtol * np.abs(ri)
+                todo[todo1] = sub_todo
+
+            if diagnostic is not None:
+                re2 = re.copy()
+                re2[todo1] = rei
+                diagnostic["residual"][k].append(re2)
+
+            if np.sum(todo) == 0:
+                break
+
+            # shrink previous vectors of estimates
+            fd = [fdi[sub_todo] for fdi in fd]
+
+        if jac.ndim == 0:
+            jac[...] = r
+            err[...] = re
+        elif jac.ndim == 1:
+            jac[kx] = r
+            err[kx] = re
+        else:
+            jac[(...,) + kx] = r
+            err[(...,) + kx] = re
+
+        it.iternext()
+
+    return jac, err
+
+
+def _steps(p, step, maxiter):
+    h0, factor = step
+    h = p * h0
+    if not h != 0:  # also works if p is NaN
+        h = h0
+    return h * factor ** np.arange(maxiter)
+
+
+def _derive(mode, f0, f, x, i, h, args):
+    x1 = x.copy()
+    x2 = x.copy()
+    if mode == 0:
+        x1[i] += h
+        x2[i] -= h
+        return (f(x1, *args) - f(x2, *args)) * (0.5 / h)
+    h = h * mode
+    x1[i] += h
+    x2[i] += 2 * h
+    f1 = f(x1, *args)
+    f2 = f(x2, *args)
+    return (-3 * f0 + 4 * f1 - f2) * (0.5 / h)
+
+
+def _first(method, f0, f, x, i, h, args):
```

Comment:
I agree that sounds like a cleaner design, but I remember I chose this design to avoid a duplicate calculation.

Reply:
I figured. I'd just suggest that there are other ways to avoid the duplicate calculation. I will probably work on that in my PR and then I'll have something more concrete to say.